learn/shorts/functional-programming.qmd

```diff
@@ -0,0 +1,170 @@
+
+    **Syntax:**
+    ```python
+    reduce(function, iterable)
+    ```
+    The `function` takes two arguments and is applied cumulatively to the elements of `iterable` to reduce them to a single value.
+    
+    It is slightly different from the first two functions in that it isn't built into Python but is available in the `functools` module.
+    
+    For example:
+    
+    ```python
+    from functools import reduce
+    numbers = [1, 2, 3, 4, 5]
+    sum_of_numbers = reduce(lambda x, y: x + y, numbers)
+    print(sum_of_numbers)
+    ```
+
+    In this final example, the `reduce` function is used to calculate the sum of the numbers in the list. The lambda function defines how the elements of the list should be combined, in this case by adding them together. As the `reduce` function returns a single value, we can print it directly.
+
+### What does the lambda keyword do in Python?
+
+The `lambda` keyword in Python is used to create what is called anonymous functions (i.e. not named, and therefore bound to the place where they are defined). In functional programming, the `lambda` function is often used to define simple functions that are used only once.
+
+**Syntax:**
+```python
+lambda arguments: expression
+```   
+A `lambda` function can take any number of arguments but must have only a single expression, which is evaluated and returned.
+
+In the `filter` example above, the lambda function (`lambda x: x % 2 == 0`) is used to define the condition for filtering the list of numbers. The argument `x` represents each element of the list, and the expression `x % 2 == 0` checks if the element is even. The lambda function used in the `map` example (`lambda x: x ** 2`) is used in much the same way, with a single argument and a simple expression.
+
+In the `reduce` example, the lambda function (`lambda x, y: x + y`) takes two arguments and returns their sum. The first argument represents the accumulated value, and the second argument represents the next element of the list. The lambda function is applied to each pair of elements in the list until a single value is returned.
+
+### What is list comprehension in Python?
+
+The final concept we will cover is list comprehension, which is a concise way to create lists in Python. List comprehension allows you to create a new list by applying an expression to each element of an existing list.
+
+**Syntax:**
+```python
+[expression for item in iterable]
+```
+An optional `if` condition can be added:
+```python
+[expression for item in iterable if condition]
+```
+
+For example:
+
+```python
+numbers = [1, 2, 3, 4, 5]
+squared_numbers = [x ** 2 for x in numbers]
+print(squared_numbers)
+```
+
+In this example, the list comprehension `[x ** 2 for x in numbers]` is used to create a new list of squared numbers from the list of numbers. The expression `x ** 2` is applied to each element of the list, and the resulting list is stored in the variable `squared_numbers`.
```

Comment:
This doesn't really belong here. List comprehensions are kinda a Python way of implementing a for loop that takes some of the advantages of things like `map` but keeping the syntax of a for loop. Maybe include list comprehensives as a note of a similar but alternative way to doing map without having to lose familiarity with for loops.